tripy/tripy/frontend/utils.py

```diff
@@ -212,9 +240,30 @@ def impl(func):
         @functools.wraps(func)
         def wrapper(*args, **kwargs):
             from tripy.frontend.tensor import Tensor
+            from tripy.frontend.shape import Shape, ShapeScalar
 
             all_args = utils.merge_function_arguments(func, *args, **kwargs)
 
+            # TODO: Should this be done a different way? Maybe we should manually specify which
+            # arguments to check across, similar to sync_arg_types? This could break in cases where
+            # we have operations that specifically require different tensor types.
```

Comment:
Several open questions here:
1. Should there be a parameter that controls which inputs we apply this logic across?
2. Does the hierarchy even make sense?
3. How do we ensure the behavior of ops is also consistent with this hierarchy? Ops are slightly different in that the types of the inputs should determine the type of the output, while in this decorator we only care about the inputs.

Reply:
Very curious issue to ponder. Per [my comment](https://github.com/NVIDIA/TensorRT-Incubator/issues/226#issuecomment-2377909735) in #226, I think putting Shape and Tensor in a hierarchy in this manner might not be advisable. It might cause fewer surprises to require manual conversions between these types where it's ambiguous, though `ShapeScalar` does not present this sort of problem.

Perhaps we could have an API by which operator implementations could check what types they do or do not accept for certain arguments?

Reply:
We could also take the rank into account, since shapes must be of rank 1 and shape scalars must be of rank 0.

Reply:
@slyubomirsky I'm not quite sure what the best mechanism would be to disallow mixing Shape/Tensor in these APIs. I've filed an issue for now: https://github.com/NVIDIA/TensorRT-Incubator/issues/233

In the meantime, I think this hierarchy works for all the ops that we currently have. Maybe we can revisit this if/when it breaks in the future? I would like to check this in so we will finally have a (mostly) passing L1.

Reply:
If we want to have a hierarchy as a stopgap measure, that's fine. However, we should make it a lattice in that case.
```
             Tensor
            /         \
        Shape   ShapeScalar
```
All `Shape`s are `Tensor`s and all `ShapeScalar`s are `Tensor`s, but `ShapeScalar`s are not `Shape`s.

Reply:
But when you have `Shape <op> ShapeScalar`, you would generally want a `Shape` back, since the scalar will be broadcasted. The priority ordering here does not imply anything about the inheritance hierarchy which is indeed the lattice you describe.

Reply:
Ah that's true, I hadn't taken into account broadcasting. That works, in that case.